hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/node/DatanodeAdminMonitorImpl.java

```diff
@@ -238,10 +295,11 @@ private void processPendingNodes() {
 
   private void processTransitioningNodes() {
```

Comment:
The current code in `DatanodeAdminMonitorImpl` that reads and writes to the set `trackedNodes` is not thread safe. In this method, a thread is iterating through `trackedNodes` and can modify it by using the iterator. Meanwhile, another thread could read this set in the `getTrackedNodes` and `getSingleTrackedNode` methods. The latter two methods are synchronized, but `processTransitioningNodes` is not, which is incorrect. We could fix this by making this method synchronized. It's also possible to do fine grained synchronization by wrapping the `iterator.remove()` calls with synchronized. What do you think?

Reply:
What about just changing `trackedNodes` to be a concurrent Hash set `ConcurrentHashMap.newKeySet();`? That would avoid a potential problem of `processTransitioningNodes` taking a long time and lock out calls to the other methods while it is running.

Reply:
Yeah, a concurrent set should also work well enough for our purpose I think.

Reply:
I have changed it to ConcurrentHashMap.newKeySet() in the recent commits. Could you check if it is fine now?

Reply:
Looks fine. I've kicked off CI.